File: api/src/lexical_search.py
import argparse
import hashlib
import logging
import pathlib
import re
import sqlite3
import unicodedata
from dataclasses import dataclass
from typing import Any, Dict, Iterable, List


logger = logging.getLogger(__name__)

SCRIPT_DIR = pathlib.Path(__file__).parent.parent
TOKEN_RE = re.compile(r"[a-z0-9]{2,}")


@dataclass(frozen=True)
class LexicalCorpusConfig:
    name: str
    source_root: pathlib.Path
    file_glob: str
    db_path: pathlib.Path


TECH_DOCS_LEXICAL_CONFIG = LexicalCorpusConfig(
    name="tech_docs",
    source_root=SCRIPT_DIR / "data" / "a220-tech-docs" / "ocr",
    file_glob="*.md",
    db_path=SCRIPT_DIR / "data" / "a220-tech-docs" / "lexical" / "fts.sqlite3",
)

NC_LEXICAL_CONFIG = LexicalCorpusConfig(
    name="non_conformities",
    source_root=SCRIPT_DIR / "data" / "a220-non-conformities" / "md",
    file_glob="*.md",
    db_path=SCRIPT_DIR / "data" / "a220-non-conformities" / "lexical" / "fts.sqlite3",
)

DEFAULT_LEXICAL_CORPORA = {
    TECH_DOCS_LEXICAL_CONFIG.name: TECH_DOCS_LEXICAL_CONFIG,
    NC_LEXICAL_CONFIG.name: NC_LEXICAL_CONFIG,
}


def normalize_text(value: str) -> str:
    return (
        unicodedata.normalize("NFKD", value)
        .encode("ascii", "ignore")
        .decode("ascii")
        .lower()
    )


def tokenize_query(value: str) -> List[str]:
    deduped: List[str] = []
    seen = set()
    for token in TOKEN_RE.findall(normalize_text(value)):
        if token in seen:
            continue
        seen.add(token)
        deduped.append(token)
    return deduped


def build_match_query(value: str, *, operator: str = "AND") -> str:
    tokens = tokenize_query(value)
    if not tokens:
        return ""
    joiner = f" {operator} "
    return joiner.join(f"{token}*" for token in tokens)


def iter_corpus_files(config: LexicalCorpusConfig) -> List[pathlib.Path]:
    return sorted(config.source_root.glob(config.file_glob))


def compute_corpus_fingerprint(paths: Iterable[pathlib.Path]) -> Dict[str, Any]:
    digest = hashlib.sha256()
    count = 0
    for path in paths:
        stat = path.stat()
        digest.update(str(path).encode("utf-8"))
        digest.update(str(stat.st_size).encode("utf-8"))
        digest.update(str(stat.st_mtime_ns).encode("utf-8"))
        count += 1
    return {
        "document_count": count,
        "fingerprint": digest.hexdigest(),
    }


def connect_fts(db_path: pathlib.Path) -> sqlite3.Connection:
    db_path.parent.mkdir(parents=True, exist_ok=True)
    connection = sqlite3.connect(db_path)
    connection.row_factory = sqlite3.Row
    return connection


def ensure_schema(connection: sqlite3.Connection) -> None:
    connection.execute(
        """
        CREATE TABLE IF NOT EXISTS lexical_meta (
            key TEXT PRIMARY KEY,
            value TEXT NOT NULL
        )
        """
    )
    connection.execute(
        """
        CREATE VIRTUAL TABLE IF NOT EXISTS lexical_documents
        USING fts5(
            doc,
            chunk_id,
            content,
            source_path UNINDEXED,
            tokenize = 'unicode61 remove_diacritics 2'
        )
        """
    )


def read_meta(connection: sqlite3.Connection, key: str) -> str | None:
    row = connection.execute(
        "SELECT value FROM lexical_meta WHERE key = ?",
        (key,),
    ).fetchone()
    return None if row is None else row["value"]


def write_meta(connection: sqlite3.Connection, key: str, value: str) -> None:
    connection.execute(
        """
        INSERT INTO lexical_meta(key, value)
        VALUES(?, ?)
        ON CONFLICT(key) DO UPDATE SET value = excluded.value
        """,
        (key, value),
    )


def rebuild_lexical_index(
    config: LexicalCorpusConfig,
    *,
    force: bool = False,
) -> Dict[str, Any]:
    paths = iter_corpus_files(config)
    if not paths:
        raise FileNotFoundError(
            f"No source files found for lexical corpus '{config.name}' in {config.source_root}"
        )

    fingerprint_info = compute_corpus_fingerprint(paths)
    connection = connect_fts(config.db_path)
    ensure_schema(connection)

    existing_fingerprint = read_meta(connection, "fingerprint")
    should_rebuild = force or existing_fingerprint != fingerprint_info["fingerprint"]

    if should_rebuild:
        logger.info("Rebuilding lexical index for %s at %s", config.name, config.db_path)
        with connection:
            connection.execute("DELETE FROM lexical_documents")
            for path in paths:
                connection.execute(
                    """
                    INSERT INTO lexical_documents(doc, chunk_id, content, source_path)
                    VALUES(?, ?, ?, ?)
                    """,
                    (
                        path.name,
                        path.stem,
                        path.read_text(encoding="utf-8", errors="ignore"),
                        str(path),
                    ),
                )
            write_meta(connection, "fingerprint", fingerprint_info["fingerprint"])
            write_meta(connection, "document_count", str(fingerprint_info["document_count"]))
    else:
        logger.info("Lexical index for %s is already up to date", config.name)

    row_count = connection.execute(
        "SELECT COUNT(*) AS count FROM lexical_documents"
    ).fetchone()["count"]
    connection.close()

    return {
        "corpus": config.name,
        "db_path": str(config.db_path),
        "source_root": str(config.source_root),
        "document_count": row_count,
        "fingerprint": fingerprint_info["fingerprint"],
        "rebuilt": should_rebuild,
    }


def ensure_lexical_index_exists(config: LexicalCorpusConfig) -> Dict[str, Any]:
    if not config.db_path.exists():
        return rebuild_lexical_index(config)

    connection = connect_fts(config.db_path)
    ensure_schema(connection)
    row_count = connection.execute(
        "SELECT COUNT(*) AS count FROM lexical_documents"
    ).fetchone()["count"]
    fingerprint = read_meta(connection, "fingerprint")
    connection.close()

    return {
        "corpus": config.name,
        "db_path": str(config.db_path),
        "source_root": str(config.source_root),
        "document_count": row_count,
        "fingerprint": fingerprint,
        "rebuilt": False,
    }


def search_lexical_corpus(
    config: LexicalCorpusConfig,
    query: str,
    *,
    limit: int = 10,
) -> List[Dict[str, Any]]:
    ensure_summary = ensure_lexical_index_exists(config)
    match_query = build_match_query(query, operator="AND")

    if not match_query:
        logger.info("Lexical query is empty after normalization for %s", config.name)
        return []

    connection = connect_fts(config.db_path)
    ensure_schema(connection)
    def fetch_rows(current_match_query: str) -> List[sqlite3.Row]:
        return connection.execute(
        """
        SELECT
            doc,
            chunk_id,
            content,
            source_path,
            bm25(lexical_documents, 8.0, 4.0, 1.0) AS bm25_score
        FROM lexical_documents
        WHERE lexical_documents MATCH ?
        ORDER BY bm25_score ASC, doc ASC
        LIMIT ?
        """,
            (current_match_query, limit),
        ).fetchall()

    rows = fetch_rows(match_query)
    if not rows and len(tokenize_query(query)) > 1:
        match_query = build_match_query(query, operator="OR")
        rows = fetch_rows(match_query)
    connection.close()

    results: List[Dict[str, Any]] = []
    for rank, row in enumerate(rows, start=1):
        results.append(
            {
                "doc": row["doc"],
                "chunk_id": row["chunk_id"],
                "content": row["content"],
                "source_path": row["source_path"],
                "match_query": match_query,
                "bm25_score": row["bm25_score"],
                "lexical_rank": rank,
                "corpus": config.name,
                "index_document_count": ensure_summary["document_count"],
            }
        )
    return results


def rebuild_default_lexical_indexes(force: bool = False) -> List[Dict[str, Any]]:
    return [
        rebuild_lexical_index(config, force=force)
        for config in DEFAULT_LEXICAL_CORPORA.values()
    ]


def search_documents_lexical(query: str, n_results: int = 10) -> List[Dict[str, Any]]:
    return search_lexical_corpus(TECH_DOCS_LEXICAL_CONFIG, query, limit=n_results)


def search_non_conformities_lexical(query: str, n_results: int = 10) -> List[Dict[str, Any]]:
    return search_lexical_corpus(NC_LEXICAL_CONFIG, query, limit=n_results)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Build or query lexical SQLite FTS5 indexes.")
    parser.add_argument(
        "--corpus",
        choices=["all", *DEFAULT_LEXICAL_CORPORA.keys()],
        default="all",
        help="Corpus to build/query.",
    )
    parser.add_argument(
        "--force",
        action="store_true",
        help="Force a rebuild even if the fingerprint did not change.",
    )
    parser.add_argument(
        "--query",
        default="",
        help="Optional lexical query to run after index creation.",
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=5,
        help="Maximum number of query hits to print.",
    )
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    corpus_names = (
        list(DEFAULT_LEXICAL_CORPORA.keys())
        if args.corpus == "all"
        else [args.corpus]
    )

    summaries = []
    for corpus_name in corpus_names:
        config = DEFAULT_LEXICAL_CORPORA[corpus_name]
        summary = rebuild_lexical_index(config, force=args.force)
        summaries.append(summary)
        print(summary)
        if args.query:
            hits = search_lexical_corpus(config, args.query, limit=args.limit)
            print({"corpus": corpus_name, "query": args.query, "hits": hits})


if __name__ == "__main__":
    main()

File: backend-ts/src/services/lexical-search.ts
import { existsSync, mkdirSync, readdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { DatabaseSync } from "node:sqlite";

const API_ROOT = fileURLToPath(new URL("../../../api/", import.meta.url));
const TOKEN_RE = /[a-z0-9]{2,}/g;

export interface LexicalCorpusConfig {
  readonly name: string;
  readonly sourceRoot: string;
  readonly fileGlobSuffix: string;
  readonly dbPath: string;
}

export const TECH_DOCS_LEXICAL_CONFIG: LexicalCorpusConfig = {
  name: "tech_docs",
  sourceRoot: path.join(API_ROOT, "data", "a220-tech-docs", "ocr"),
  fileGlobSuffix: ".md",
  dbPath: path.join(API_ROOT, "data", "a220-tech-docs", "lexical", "fts.sqlite3"),
};

export const NC_LEXICAL_CONFIG: LexicalCorpusConfig = {
  name: "non_conformities",
  sourceRoot: path.join(API_ROOT, "data", "a220-non-conformities", "md"),
  fileGlobSuffix: ".md",
  dbPath: path.join(API_ROOT, "data", "a220-non-conformities", "lexical", "fts.sqlite3"),
};

export function normalizeText(value: string): string {
  return value
    .normalize("NFKD")
    .replace(/[^\x00-\x7F]/g, "")
    .toLowerCase();
}

export function tokenizeQuery(value: string): string[] {
  const deduped: string[] = [];
  const seen = new Set<string>();
  const matches = normalizeText(value).match(TOKEN_RE) ?? [];
  for (const token of matches) {
    if (seen.has(token)) {
      continue;
    }
    seen.add(token);
    deduped.push(token);
  }
  return deduped;
}

export function buildMatchQuery(value: string, operator: "AND" | "OR" = "AND"): string {
  const tokens = tokenizeQuery(value);
  if (tokens.length === 0) {
    return "";
  }
  return tokens.map((token) => `${token}*`).join(` ${operator} `);
}

function connectFts(dbPath: string): DatabaseSync {
  mkdirSync(path.dirname(dbPath), { recursive: true });
  return new DatabaseSync(dbPath);
}

function ensureSchema(db: DatabaseSync): void {
  db.exec(`
    CREATE TABLE IF NOT EXISTS lexical_meta (
      key TEXT PRIMARY KEY,
      value TEXT NOT NULL
    );
    CREATE VIRTUAL TABLE IF NOT EXISTS lexical_documents
    USING fts5(
      doc,
      chunk_id,
      content,
      source_path UNINDEXED,
      tokenize='unicode61 remove_diacritics 2'
    );
  `);
}

function iterCorpusFiles(config: LexicalCorpusConfig): string[] {
  return readdirSync(config.sourceRoot)
    .filter((filename) => filename.endsWith(config.fileGlobSuffix))
    .sort()
    .map((filename) => path.join(config.sourceRoot, filename));
}

function computeCorpusFingerprint(paths: readonly string[]): {
  readonly documentCount: number;
  readonly fingerprint: string;
} {
  const parts: string[] = [];
  for (const filePath of paths) {
    const stats = statSync(filePath);
    parts.push(filePath, String(stats.size), String(stats.mtimeMs));
  }
  return {
    documentCount: paths.length,
    fingerprint: normalizeText(parts.join("|")),
  };
}

function readMeta(db: DatabaseSync, key: string): string | null {
  const row = db.prepare(`SELECT value FROM lexical_meta WHERE key = ?`).get(key) as
    | { value: string }
    | undefined;
  return row?.value ?? null;
}

function writeMeta(db: DatabaseSync, key: string, value: string): void {
  db.prepare(`
    INSERT INTO lexical_meta(key, value)
    VALUES(?, ?)
    ON CONFLICT(key) DO UPDATE SET value = excluded.value
  `).run(key, value);
}

export function rebuildLexicalIndex(
  config: LexicalCorpusConfig,
  options: { readonly force?: boolean } = {},
): {
  readonly corpus: string;
  readonly dbPath: string;
  readonly sourceRoot: string;
  readonly documentCount: number;
  readonly fingerprint: string;
  readonly rebuilt: boolean;
} {
  const paths = iterCorpusFiles(config);
  if (paths.length === 0) {
    throw new Error(`No source files found for lexical corpus '${config.name}' in ${config.sourceRoot}`);
  }

  const fingerprintInfo = computeCorpusFingerprint(paths);
  const db = connectFts(config.dbPath);
  ensureSchema(db);
  const existingFingerprint = readMeta(db, "fingerprint");
  const shouldRebuild = options.force === true || existingFingerprint !== fingerprintInfo.fingerprint;

  if (shouldRebuild) {
    db.exec(`DELETE FROM lexical_documents;`);
    const insert = db.prepare(`
      INSERT INTO lexical_documents(doc, chunk_id, content, source_path)
      VALUES(?, ?, ?, ?)
    `);
    for (const filePath of paths) {
      insert.run(
        path.basename(filePath),
        path.basename(filePath, path.extname(filePath)),
        readFileSync(filePath, "utf8"),
        filePath,
      );
    }
    writeMeta(db, "fingerprint", fingerprintInfo.fingerprint);
    writeMeta(db, "document_count", String(fingerprintInfo.documentCount));
  }

  const row = db.prepare(`SELECT COUNT(*) AS count FROM lexical_documents`).get() as { count: number };
  db.close();
  return {
    corpus: config.name,
    dbPath: config.dbPath,
    sourceRoot: config.sourceRoot,
    documentCount: row.count,
    fingerprint: fingerprintInfo.fingerprint,
    rebuilt: shouldRebuild,
  };
}

export function ensureLexicalIndexExists(config: LexicalCorpusConfig): {
  readonly corpus: string;
  readonly dbPath: string;
  readonly sourceRoot: string;
  readonly documentCount: number;
  readonly fingerprint: string | null;
  readonly rebuilt: boolean;
} {
  const db = connectFts(config.dbPath);
  ensureSchema(db);
  const row = db.prepare(`SELECT COUNT(*) AS count FROM lexical_documents`).get() as { count: number };
  const fingerprint = readMeta(db, "fingerprint");
  db.close();

  if (row.count === 0 && !existsSync(config.sourceRoot)) {
    return {
      corpus: config.name,
      dbPath: config.dbPath,
      sourceRoot: config.sourceRoot,
      documentCount: 0,
      fingerprint,
      rebuilt: false,
    };
  }

  if (row.count === 0) {
    return rebuildLexicalIndex(config);
  }

  return {
    corpus: config.name,
    dbPath: config.dbPath,
    sourceRoot: config.sourceRoot,
    documentCount: row.count,
    fingerprint,
    rebuilt: false,
  };
}

export interface LexicalSearchResult extends Record<string, unknown> {
  readonly doc: string;
  readonly chunk_id: string;
  readonly content: string;
  readonly source_path: string;
  readonly match_query: string;
  readonly bm25_score: number;
  readonly lexical_rank: number;
  readonly corpus: string;
  readonly index_document_count: number;
}

export function searchLexicalCorpus(
  config: LexicalCorpusConfig,
  query: string,
  options: { readonly limit?: number } = {},
): LexicalSearchResult[] {
  const summary = ensureLexicalIndexExists(config);
  let matchQuery = buildMatchQuery(query, "AND");
  if (!matchQuery) {
    return [];
  }

  const db = connectFts(config.dbPath);
  ensureSchema(db);
  const fetchRows = (currentMatchQuery: string): Array<Record<string, unknown>> =>
    db.prepare(`
      SELECT
        doc,
        chunk_id,
        content,
        source_path,
        bm25(lexical_documents, 8.0, 4.0, 1.0) AS bm25_score
      FROM lexical_documents
      WHERE lexical_documents MATCH ?
      ORDER BY bm25_score ASC, doc ASC
      LIMIT ?
    `).all(currentMatchQuery, options.limit ?? 10) as Array<Record<string, unknown>>;

  let rows = fetchRows(matchQuery);
  if (rows.length === 0 && tokenizeQuery(query).length > 1) {
    matchQuery = buildMatchQuery(query, "OR");
    rows = fetchRows(matchQuery);
  }
  db.close();

  return rows.map((row, index) => ({
    doc: String(row.doc),
    chunk_id: String(row.chunk_id),
    content: String(row.content),
    source_path: String(row.source_path),
    match_query: matchQuery,
    bm25_score: Number(row.bm25_score),
    lexical_rank: index + 1,
    corpus: config.name,
    index_document_count: summary.documentCount,
  }));
}

export function searchDocumentsLexical(query: string, limit = 10): LexicalSearchResult[] {
  return searchLexicalCorpus(TECH_DOCS_LEXICAL_CONFIG, query, { limit });
}

export function searchNonConformitiesLexical(query: string, limit = 10): LexicalSearchResult[] {
  return searchLexicalCorpus(NC_LEXICAL_CONFIG, query, { limit });
}
